### ISSUE TITLE
Missing attention mask for input images

### ISSUE THREAD

**Issue by AmitMY** [sign/image-latent-transformer#1]
When collating different words, their image size is variable.
Currently, it is 0-padded, but this is not something the model should care about.
Instead, we need to add an attention mask for the pixels as well, to prevent the model from behaving differently when the text is batched with longer or shorter texts

**Comment by AmitMY** [comment]
Huggingface issue: https://discuss.huggingface.co/t/image-encoder-inference-with-variable-size-imags/166910

**Comment by AmitMY** [comment]
Addressed the determinism issue in https://github.com/sign/image-latent-transformer/commit/0f4164567c57089e54d2a239182fdd46515bd611 - but my solution is slow.

Should use parallelizable models like dino 3 to solve that issue. https://github.com/facebookresearch/dinov3/issues/27#issuecomment-3224039775

**Comment by AmitMY** [comment]
Ideally, our collator should collate all images into a tensor of `[B, L, H, W, 3]` - this will prevent current issues with data transfer (such as relieved here https://github.com/sign/image-latent-transformer/commit/c19b7e67216fa541dac2142d941dcdc4337a94e1)

Currently, we pass a list of a nested tensor.

Once we can pass a `[B, L, H, W, 3]` tensor, padded with 0, and that is deterministic (0s are ignored correctly) we can:
1. Use stack tensors instead of `torch.nested` from images https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/processor.py#L136-L138
2. Remove this type: https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/batch_image_encoder.py#L11
3. Remove this hack: https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/batch_image_encoder.py#L69-L70
4. Never encode images sequentially https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/batch_image_encoder.py#L133C5-L147
5. Remove fixed renderer size https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/renderer.py#L71-L73
6. Remove special case for collator https://github.com/sign/image-latent-transformer/blob/c19b7e67216fa541dac2142d941dcdc4337a94e1/image_latent_transformer/collator.py#L89-L92

**Comment by AmitMY** [comment]
We now have NaViT support, but it is also too slow.
Benchmark:
https://github.com/sign/image-latent-transformer/blob/main/examples/benchmark_image_encoder.py
Results (see https://github.com/lucidrains/vit-pytorch/issues/347):

1. ViT Model with padding (inconsistent results) - 1 second
2. Vit Model grouped into batches of same size images - 13 seconds
3. Vit Model patched (consistent results) - 34 second
4. NaVit Model - 39 seconds

Is the best solution to run with padding, but then "fine tune" using grouping?

**Comment by AmitMY** [comment]
next steps:
1. try to get NaViT model to be as fast as ViT - that would be very cool, but would limit the architecture of the image encoder
2. use "futures" like style, [dispatch all groups simultaneously](https://chatgpt.com/share/68e61ed8-8a90-800e-9e9c-2349d01d6aac), to try and saturate the GPU
